### DISCUSSION TITLE
Monitoramento de erros e bugs

### DISCUSSION THREAD

**Discussion by jdgabriel** [frontendbr/forum#2348] [Perguntas e Respostas]
Fala galera... blza?
Chega uma hora na vida de um `dev jr` que ele tem que aprender como fazer um monitoramento de erros e bugs que não seja pelo `console.log` (Nada contra, uso muito).

Poderia me indicar algum `recurso / blog / post` que possa me mostrar e explicar como funciona o monitoramento de erros em um projeto de grande porte?
 Estou lendo sobre o `prom-client` com `Grafana`, mas não estou conseguindo ver em como esses logs podem me ajudar? (tudo é gráfico kk)

**Comment by felquis** [comment]
Quando o site está rodando em seu computador, você consegue ver vários logs ali no terminal, e no desenvolvimento local.

Porém quando o site está rodando em produção, ele pode gerar muitos logs, muitos mesmo. No fim das contas, é um monte de log que você precisa ter em uma interface de fácil manuseio que te ajuda a entender se o sistema está rodando da maneira esperada.

Não conheço o prom-client prometheus https://prometheus.io/ e grafana https://grafana.com/ mas são soluções de telemetria open source que você mesmo vai gerenciar e manter online, mas existem SaaS que você tem apenas que usar os endpoints de uma empresa que mantem os servidores de pé para você, como datadog, sentry para citar dois bem genéricos.

O gráfico pode te ajudar a ver muitos dados em um "pico" ou "spike" onde o gráfico tem uma mudança abrupta, então as vezes o gráfico serve para você ter uma leitura mais rápida da frequência que alguns eventos ocorrem. Ex: você observa que as 8am tem um uso excessivo de CPU do banco, e descobre que neste horário tem um cronjob que faz leitura e escrita todos os registros de várias tabelas.

O gráfico ajuda muito a te dar recursos para analisar coisas que se você ler os logs linha por linha, levaria um bom tempo para entender "As 8am tem um aumento do uso do CPU" no gráfico você já entende isso de primeira. E há muitos exemplos, como:

- monitoramento de memória, cpu das VMs/computadores que rodam o sistema e dos bancos de dados.
- monitoramento de erros de runtime, tanto de front-end quanto back-end. Pode incluir tbm o monitoramento de logs dos bancos de dados.
- monitoramento dos status HTTP que a aplicação está respondendo, incluindo quanto tempo os requests estão levando para concluir.

Se você não tem uma ferramenta de telemetria dessas, você só ficaria sabendo que algo não está funcionando quando um cliente reclamasse que algo não funciona e tivesse a boa vontade de reclamar, e geralmente só reclamariam se precisassem muito do sistema. Porém, toda equipe de desenvolvimento procura saber que algo não funciona de forma proativa, antes que alguém venha dizer "algo não está funcionando", para não ficar a mercê de reclamações diretas de clientes. De forma que os clientes podem até perceber que algo deu errado, mas a equipe por trás provavelmente tem vários recursos e sabem que algo está dando errado.

Especificamente de front-end, há telemetrias que ajudam a entender como as pessoas usam o sistema, exemplo: smartlook, hotjar, que inclusive gravam sessões de usuário, que ajudam muito em pesquisas de usabilidade e UX.

E Google Analytics, que ajuda bastante a quantificar dados de UX, mensurar se as pessoas estão chegando até onde gostaríamos que chagassem.

Embora, o prometheus e grafana que você mencionou são muito mais focados em, telemetria técnica da saude geral do sistema. Quis mencionar estes dois acima, pois o fórum tbm é de front-end haha e essas já me ajudaram muito também.
